### ISSUE TITLE
Should scrypt be used in place of pbkdf2 for generated keystore files?

### ISSUE THREAD

**Issue by michaelsbradleyjr** [status-im/nim-status#236]
Here's a keystore file generated by nim-status a few minutes ago:
```json
{
   "address":"6dac2d3b6670b9a7e45d10ecb6eed4939a7c385d",
   "crypto":{
      "cipher":"aes-128-ctr",
      "cipherparams":{
         "iv":"39a047b2da9dc904ba5f3cccb4cd3ec8"
      },
      "ciphertext":"cfc8963b9f354d34d4e7a02b2766e2c0a10a823f0c54d4e3248382ecee3c011c",
      "kdf":"pbkdf2",
      "kdfparams":{
         "dklen":32,
         "c":100,
         "prf":"hmac-sha256",
         "salt":"e98045f439ee6630302e182db881beec"
      },
      "mac":"8d3b0c36522cc96272b89a958838f61a5e8241b970af9e8bea8a6879a46df524"
   },
   "id":"820e83e2-499a-fd49-9e12-0f993547640c",
   "version":3
}
```

And here's one generated by status-go (as used by status-desktop) a few minutes ago:
```json
{
   "address":"7533813bbe96455d60354d6533a75e10d9476129",
   "crypto":{
      "cipher":"aes-128-ctr",
      "ciphertext":"52ceaab472966e5392278453ccf17d2dd27c7714843b06b69c968c767d875b94",
      "cipherparams":{
         "iv":"40ab9bb30f2b49cf22c9773d45c0be3b"
      },
      "kdf":"scrypt",
      "kdfparams":{
         "dklen":32,
         "n":4096,
         "p":6,
         "r":8,
         "salt":"c63e2c31356b37eedbaadf05845d7a840ea7e8d247e02cf2e4dc558c03458d54"
      },
      "mac":"c30c9579f141285f41bd63f938c2587538ef904b708264ee6ea823f7dda71a6d"
   },
   "id":"a77022f0-a782-47a6-a41a-d4deb03c4f6e",
   "version":3,
   "extendedkey":{
      "cipher":"aes-128-ctr",
      "ciphertext":"9e448bc8fb15fceefc2c2614889955f7633d183d94081b927c65d02834477ee0a7b3b202cba2a119da2b135f816c88ab2c52c3fd69be05988f9e72f13ff6456c517350718a367c30a1cabf62eca005a4f15660220d424faa6773576785bf596fd7b11e0212fe0ae5609035aefc86fc",
      "cipherparams":{
         "iv":"85418f5d13c6aa21096cb4f0bfc918fb"
      },
      "kdf":"scrypt",
      "kdfparams":{
         "dklen":32,
         "n":4096,
         "p":6,
         "r":8,
         "salt":"301fa6a2aad13d5645fe94a32b90763097301a9e21541fcf3d3563b73be5abe0"
      },
      "mac":"9281ae2d1f33c91bbc98667eee5a1fb7b39015b58ebe7dd923c0230fc09bbda5"
   },
   "subaccountindex":0
}
```

So while status-go uses `scrypt` as the kdf, nim-status uses `pbkdf2`.

The relevant modules in nim-status (current `master`) seem to be `nim_status/extkeys/mnemonic.nim` and `nim_status/accounts/generator/generator.nim`.

Note that in the latter we have a very low `workfactor: int = 0`. The question of workfactor came up in some work @emizzle and I were doing together a few weeks ago. We wanted a small value in order for our test suite to pass quickly, but maybe we need to change the default to something higher. Looking anew at the logic there re: `workfactorFinal`, I'm not sure it's doing quite what we want it to do.

A few years ago, when reading more about Ethereum, I learned (don't recall the source) that `scrypt` is preferable because it's probably harder to crack. However, that could be wrong/outdated and there may be good reasons to prefer `pbkdf2` at the present time, i.e. if it's used with suitable parameters.

On the other hand, `scrypt` is available to us via nimcrypto. So, if in fact `scrypt` is more resistant, maybe we should make the switch.

Any thoughts @siphiuel and @richard-ramos?

**Comment by siphiuel** [comment]
Seems that indeed `scrypt` is an improvement over `pbkdf2` as it's both cpu and mem heavy. I agree that moving to `scrypt` makes sense, given that status-go uses it.

There are also some even more robust variants e.g. argon2 (nim impl: https://github.com/D-Nice/argon2_bind).

**Comment by michaelsbradleyjr** [comment]
Nice find re: `argon2`.

See: https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition

That wiki entry is rather dated, but I haven't (yet) found anything that's newer and official-looking. Only `pbkdf2 ` and `scrypt` are mentioned, so I'd have concerns about portability if we used a different kdf.

But maybe folks are already using `argon2` (or others) to create Ethereum keystore files. If you find a write-up or examples of that, please post here and we can consider it for sure!

**Comment by arnetheduck** [comment]
@cheatfate what are we using in eth2?

**Comment by cheatfate** [comment]
First of all your number of iterations is very small and insecure it should at least `100,000`.

For example `nimbus-eth2` is using `2^18` for `scrypt` (which we are using by default) and you are using `100` for `pbkdf2` and `4096` for `scrypt`.

`argon2` is not yet specified by Ethereum2, only `pbkdf2` and `scrypt`. Also need to mention that `scrypt` internally uses `pbkdf2` iterations. So `pbkdf2` itself is not insecure.

**Comment by emizzle** [comment]
@siphiuel do you know why the status-go keystore file includes an `extendedKey`? Is that something that `nim-status` needs to do as well? It seems that if both are included, then every time a keystore file is created, there would need to be two encryption runs (one for the "main" key, and one for the extended key).

**Comment by siphiuel** [comment]
@emizzle no idea, will investigate

**Comment by cammellos** [comment]
@gravityblast will know, though is on holidays this week

**Comment by gravityblast** [comment]
> @siphiuel do you know why the status-go keystore file includes an extendedKey? Is that something that nim-status needs to do as well? It seems that if both are included, then every time a keystore file is created, there would need to be two encryption runs (one for the "main" key, and one for the extended key).

@cammellos @emizzle yes currently in the same file we have the key and the extended key, which is used mostly for the BIP44 parent to be able to derive new accounts.
